File: constraints/dataAnalysis/stellarMassFunctions_ZFOURGE_z0.2_2.5/massRedshiftRelation.pl
#!/usr/bin/env perl
use strict;
use warnings;
use Cwd;
use lib exists($ENV{'GALACTICUS_ROOT_V094'}) ? $ENV{'GALACTICUS_ROOT_V094'}.'/perl' : cwd().'/perl';
use PDL;
use PDL::NiceSlice;
use PDL::Fit::Polynomial;
use PDL::IO::Misc;
use GnuPlot::PrettyPlots;
use GnuPlot::LaTeX;

# Fit stellar mass completeness limits from the ZFOURGE survey (Tomczak et al. 2014). Fits the tabulated results in the data file
# given by R. Quadri.
# Andrew Benson (11-August-2014)

# Read data.
my @fields = 
    (
     {
	 name => "NMBS"
     },
     {
	 name => "ZFOURGE"
     }
    );
(my $redshift, my $massNMBS, my $massZFOURGE) = rcols("zfourge-SMF-supplemental/masslimits.dat",0,1,2);
$fields[0]->{'mass'    } = $massNMBS;
$fields[1]->{'mass'    } = $massZFOURGE;
$fields[0]->{'redshift'} = $redshift;
$fields[1]->{'redshift'} = $redshift;

# Create a plot.
my $plot;
my $gnuPlot;
my $plotFile = "constraints/dataAnalysis/stellarMassFunctions_ZFOURGE_z0.2_2.5/massRedshiftRelation.pdf";
(my $plotFileEPS = $plotFile) =~ s/\.pdf$/.eps/;
open($gnuPlot,"|gnuplot");
print $gnuPlot "set terminal epslatex color colortext lw 2 7\n";
print $gnuPlot "set output '".$plotFileEPS."'\n";
print $gnuPlot "set title 'Redshift vs. Limiting Stellar Mass for Tomczak et al. (2014) Sample'\n";
print $gnuPlot "set xlabel 'Limiting stellar mass; \$M_\\star [M_\\odot]\$'\n";
print $gnuPlot "set ylabel 'Redshift; \$z\$'\n";
print $gnuPlot "set lmargin screen 0.15\n";
print $gnuPlot "set rmargin screen 0.95\n";
print $gnuPlot "set bmargin screen 0.15\n";
print $gnuPlot "set tmargin screen 0.95\n";
print $gnuPlot "set key spacing 1.2\n";
print $gnuPlot "set key at screen 0.74,0.16\n";
print $gnuPlot "set key left\n";
print $gnuPlot "set key bottom\n";
print $gnuPlot "set logscale x\n";
print $gnuPlot "set mxtics 10\n";
print $gnuPlot "set format x '\$10^{\%L}\$'\n";
print $gnuPlot "set xrange [3.0e7:3.0e11]\n";
print $gnuPlot "set yrange [0.2:3.1]\n";
print $gnuPlot "set pointsize 2.0\n";

# Fit limiting redshifts.
my $iField = -1;
foreach my $field ( @fields ) {
    ++$iField;
    (my $fit, my $coeffs) = fitpoly1d($field->{'mass'},$field->{'redshift'},4);
    my $fitMass     = sequence(1000)*5.0/999.0+7.0;
    my $fitRedshift = pdl zeroes(nelem($fitMass));
    my $closing     = "";
    print $field->{'name'}." : ";
    for(my $i=0;$i<nelem($coeffs);++$i) {
	$fitRedshift += $coeffs->(($i))*($fitMass**$i);
	print $coeffs->(($i))."d0";
	if ( $i == nelem($coeffs)-1 ) {
	    print $closing."\n";
	} else {
	    print "+logarithmicMass*(";
	    $closing .= ")";
	}
    }
    &GnuPlot::PrettyPlots::Prepare_Dataset(
	\$plot,
	10.0**$fitMass,
	$fitRedshift,
	style      => "line",
	weight     => [3,1],
	color      => $GnuPlot::PrettyPlots::colorPairs{${$GnuPlot::PrettyPlots::colorPairSequences{'sequence1'}}[$iField]},
	title      => $field->{'name'}.' [fit]'
    );
    &GnuPlot::PrettyPlots::Prepare_Dataset(
	\$plot,
	10.0**$field->{'mass'},
	$field->{'redshift'},
	style      => "line",
	weight     => [5,3],
	linePattern => 3,
	color      => $GnuPlot::PrettyPlots::colorPairs{${$GnuPlot::PrettyPlots::colorPairSequences{'sequence1'}}[$iField]},
	title      => $field->{'name'}.' [observed]'
	);    
}
&GnuPlot::PrettyPlots::Plot_Datasets($gnuPlot,\$plot);
close($gnuPlot);
&GnuPlot::LaTeX::GnuPlot2PDF($plotFileEPS);

exit;

File: constraints/dataAnalysis/stellarMassFunctions_ZFOURGE_z0.2_2.5/massRedshiftRelation.py
#!/usr/bin/env python3
# Fit stellar mass completeness limits from the ZFOURGE survey (Tomczak et al. 2014). Fits the tabulated results in the data file
# given by R. Quadri.
# Andrew Benson (11-August-2014)

import os
import sys
import numpy as np
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt

execPath = os.environ.get('GALACTICUS_EXEC_PATH', '')

# Read data.
fields = [
    {'name': 'NMBS'    },
    {'name': 'ZFOURGE' },
]
data = np.loadtxt('constraints/dataAnalysis/stellarMassFunctions_ZFOURGE_z0.2_2.5/zfourge-SMF-supplemental/masslimits.dat', usecols=(0, 1, 2))
redshift     = data[:, 0]
massNMBS     = data[:, 1]
massZFOURGE  = data[:, 2]
fields[0]['mass'    ] = massNMBS
fields[1]['mass'    ] = massZFOURGE
fields[0]['redshift'] = redshift
fields[1]['redshift'] = redshift

# Create a plot.
fig, ax = plt.subplots(figsize=(8, 6))
ax.set_title(r"Redshift vs. Limiting Stellar Mass for Tomczak et al. (2014) Sample")
ax.set_xlabel(r'Limiting stellar mass; $M_\star\;[M_\odot]$')
ax.set_ylabel(r'Redshift; $z$')
ax.set_xscale('log')
ax.set_xlim(3.0e7, 3.0e11)
ax.set_ylim(0.2, 3.1)

colors = ['steelblue', 'darkorange']
for iField, field in enumerate(fields):
    # Fit a degree-4 polynomial.
    coeffsDescending = np.polyfit(field['mass'], field['redshift'], 4)
    coeffsAscending  = coeffsDescending[::-1]
    fitMass = np.linspace(7.0, 12.0, 1000)
    fitRedshift = np.zeros(len(fitMass))
    closing = ''
    print(field['name'] + ' : ', end='')
    for i, c in enumerate(coeffsAscending):
        fitRedshift += c * fitMass ** i
        print(f'{c:.10e}d0', end='')
        if i == len(coeffsAscending) - 1:
            print(closing)
        else:
            print('+logarithmicMass*(', end='')
            closing += ')'
    ax.plot(10.0 ** fitMass, fitRedshift,
            color=colors[iField], linewidth=1.5, label=field['name'] + ' [fit]')
    ax.plot(10.0 ** field['mass'], field['redshift'], '--',
            color=colors[iField], linewidth=2.5, label=field['name'] + ' [observed]')

ax.legend(loc='lower right')
plt.tight_layout()
plotFile = 'constraints/dataAnalysis/stellarMassFunctions_ZFOURGE_z0.2_2.5/massRedshiftRelation.pdf'
plt.savefig(plotFile)
plt.close()
